src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,187 @@
+}
+
+static pid_t
+spawn_cockpit_ws (const char *ws_path, int fd,
+                  struct instance_type sock_instance)
+{
+  pid_t pid = fork ();
+
+  if (pid == -1)
+    {
+      err (1, "fork() failed");
+    }
+  else if (pid > 0)
+    {
+      return pid;
+    }
+  else
+    {
+      const char *args[MAX_COCKPIT_WS_ARGS + 1];
+      int i = 0;
+      static char *pid_str;
+
+      /* pass the socket to ws like systemd activation does, see sd_listen_fds(3) */
+      if (dup2 (fd, SD_LISTEN_FDS_START) < 0)
+        err (1, "dup2() failed");
+
+      setenv ("LISTEN_FDS", "1", 1);
+      asprintfx (&pid_str, "%i", getpid ());
+      setenv ("LISTEN_PID", pid_str, 1);
+      setenv ("LISTEN_FDNAMES", instance_types[i].sockname, 1);
+
+      args[0] = ws_path;
+      for (i = 0; sock_instance.argv[i] != NULL; i++)
+        args[i + 1] = sock_instance.argv[i];
+      args[i + 1] = NULL;
+
+      execv (ws_path, (char **) args);
+
+      err (1, "spawning cockpit-ws instance failed");
+    }
+}
+
+int
+main (int argc, char *argv[])
+{
+  if (argc != 3)
+    errx (1, "Usage: socket-activation-helper $WS_PATH $SOCKETS_DIR");
+
+  const char *ws_path = argv[1];
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    ufds[i].fd = -1;
+
+  /* Set up signal handler for when cockpit-ws has exited */
+  struct sigaction sa;
+  sigemptyset (&sa.sa_mask);
+  sa.sa_flags = 0;
+  sa.sa_handler = handle_cockpit_ws_exited;
+  sigaction (SIGCHLD, &sa, NULL);
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    {
+      struct sockaddr_un addr = {
+        .sun_family = AF_UNIX
+      };
+      char *socket_path;
+      int listen_fd;
+
+      socket_to_pid[i] = 0;
+
+      asprintfx (&socket_path, "%s/%s", argv[2], instance_types[i].sockname);
+      strcpy (addr.sun_path, socket_path);
+
+      if (unlink (addr.sun_path) < 0 && errno != ENOENT)
+        err (1, "unlink() failed");
+
+      /* create a listening socket for each cockpit-ws mode */
+      listen_fd = socket (AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
+      if (listen_fd < 0)
+        err (1, "socket() failed");
+
+      if (bind (listen_fd, (struct sockaddr *) &addr, sizeof addr) < 0)
+        err (1, "%s: bind() failed", addr.sun_path);
+
+      if (listen (listen_fd, 32) < 0)
+        err (1, "%s: listen() failed", addr.sun_path);
+
+      ufds[i].fd = listen_fd;
+      ufds[i].events = POLLIN;
+    }
+
+  /* Loop waiting for incoming connects on any of the connected sockets. */
+  do
+    {
+      int rv;
+      do
+        rv = poll (ufds, N_ELEMENTS (instance_types), -1);
+      while (rv == -1 && errno == EINTR);
+
+      if (rv == -1)
+        {
+          err (1, "poll() failed");
+        }
+      else
+        {
+          for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+            {
+              if (ufds[i].revents == POLLIN)
+                {
+                  ufds[i].events = 0;
+                  socket_to_pid[i] = spawn_cockpit_ws (ws_path, ufds[i].fd,
```

Comment:
as mentioned on IRC, we could actually just mask/unmask `SIGCHLD` on either side of this call. that would ensure that the pid lands in the array before any signal handler could possibly run.

if this is just a testcase, i think it's fine not to do that, though.